Implemented request-body data insertion through text-editor

devblin opened the pull request:
Implements #41 

For methods with request body, this update now allows user to insert request body data through editor.
For example:
- Currently, this is what we do to add request body:
`$ hopp-cli post https://reqres.in/api/users/2 -c js -b '{"name": "morp","job": "zion resident"}'`
- Now, we can also use **-e** flag, indicating request data insertion through text-editor:
`$ hopp-cli post https://reqres.in/api/users/2 -c js -e`

gbmor commented on methods/basic.go:
```diff
@@ -15,7 +20,62 @@ func BasicRequestWithBody(c *cli.Context, method string) (string, error) {
+		}
+
+		// create temp file for writing request body
+		tempFile, err := ioutil.TempFile("", "HOPP-CLI-REQUEST*.txt")
```
Idiomatically, in most cases the directory prefix for temp files should be passed as `os.TempDir()` instead of an empty string. This is cross-platform.

devblin replied:
Actually, I checked **ioutil** package doc, for `ioutil.TempFile(dir, prefix)`, it says if **dir** is empty string, the default dir for temp files will be used.

gbmor replied:
Oh, woops! I've always used it explicitly, so I didn't remember that!

gbmor commented on methods/basic.go:
```diff
@@ -15,7 +20,62 @@ func BasicRequestWithBody(c *cli.Context, method string) (string, error) {
+		if err != nil {
+			return "", fmt.Errorf("Unable to create temp file: %s\n%s", tempFile.Name(), err)
+		}
+
```
`defer os.Remove(tempFile.Name())` should move to just after the error check. Otherwise, if an error condition is reached before the deferred deletion, you'll pollute the filesystem with empty temp files.

The convention in Go is to place cleanup `defer` statements immediately after the error check.

gbmor commented on methods/basic.go:
```diff
@@ -15,7 +20,62 @@ func BasicRequestWithBody(c *cli.Context, method string) (string, error) {
+		cmd.Wait()
+
+		// read temp file contents
+		fileData, err := ioutil.ReadFile(tempFileName)
+		if err != nil {
+			return "", fmt.Errorf("Error reading file: %s", err)
+		}
+
+		jsonStr = []byte(string(fileData))
```
this is a conversion from []byte->string->[]byte, which isn't needed. Since `err` has already been declared as well as `jsonStr`, you can assign directly without the initializer shorthand, like:

```go
jsonStr, err = ioutil.ReadFile(tempFileName)
```

gbmor commented on methods/basic.go:
```diff
@@ -15,7 +20,62 @@ func BasicRequestWithBody(c *cli.Context, method string) (string, error) {
+
+		jsonStr = []byte(string(fileData))
+
+		defer os.Remove(tempFileName)
```
See the above comment about the position of the deferred statement.

gbmor commented on methods/basic.go:
```diff
@@ -15,7 +20,62 @@ func BasicRequestWithBody(c *cli.Context, method string) (string, error) {
+		currentOs := runtime.GOOS
+
+		// based on OS find assign default editor
+		if currentOs == "linux" || currentOs == "darwin" {
+			editor = "nano"
```
It's better to respect `$EDITOR` on linux and macOS. You can check if it is set, and if not, default to nano. By default on, eg, Ubuntu, it is not set, so if someone has set it, then it is an explicit preference.

`os.Getenv("EDITOR")` will return an empty string if it is not set.

Something like:

```go
editor = os.Getenv("EDITOR")
if editor == "" {
    editor = "nano"
}
```

gbmor commented on methods/basic.go:
```diff
@@ -15,7 +20,62 @@ func BasicRequestWithBody(c *cli.Context, method string) (string, error) {
+
+		path, err = exec.LookPath(editor)
+		if err != nil {
+			return "", fmt.Errorf("Unable to locate %s at %s", editor, path)
```
Generally, when bubbling up errors, it is a good idea to wrap the original (Errorf format verb `%w` in go1.13+) so you get the exact reason for the failure. Exceptions exist, but since this is running on a local user's system as a cli tool, this is relevant info.

Also, if err is not nil, path will be an empty string, so including the value of `path` will just be include an empty string.

Eg:
```go
return "", fmt.Errorf("unable to locate %s: %w", editor, err)
```

gbmor commented on methods/basic.go:
```diff
@@ -15,7 +20,62 @@ func BasicRequestWithBody(c *cli.Context, method string) (string, error) {
+		// create temp file for writing request body
+		tempFile, err := ioutil.TempFile("", "HOPP-CLI-REQUEST*.txt")
+		if err != nil {
+			return "", fmt.Errorf("Unable to create temp file: %s\n%s", tempFile.Name(), err)
```
the error should be wrapped instead of stringified and not broken to a separate line.

Also, if err is not nil here, tempFile *will* be nil, so calling `tempFile.Name()` will be a runtime panic from a nil pointer dereference.

Eg:
```go
return "", fmt.Errorf("unable to create temp file: %w", err)
```

gbmor commented on methods/basic.go:
```diff
@@ -15,7 +20,62 @@ func BasicRequestWithBody(c *cli.Context, method string) (string, error) {
+
+		err = cmd.Start()
+		if err != nil {
+			return "", fmt.Errorf("Unable to open editor: %s", err)
```
Wrap this error instead of stringifying it as well:

```go
return "", fmt.Errorf("unable to open editor: %w", err)
```

gbmor commented on methods/basic.go:
```diff
@@ -15,7 +20,62 @@ func BasicRequestWithBody(c *cli.Context, method string) (string, error) {
+		// read temp file contents
+		fileData, err := ioutil.ReadFile(tempFileName)
+		if err != nil {
+			return "", fmt.Errorf("Error reading file: %s", err)
```
Wrap this error instead of stringifying it as well:

```go
return "", fmt.Errorf("error reading file: %w", err)
```

gbmor reviewed (commented):
There are some idioms that were missed that are typically used in Go, a possibility for filesystem pollution with stray files, a possible runtime panic, some extra allocations (from redundant type conversion), and one usability concern.

The only major thing is the runtime panic, and maybe the stray files. the rest are just suggestions for how things are typically used in Go. :smile:

gbmor commented on methods/basic.go:
```diff
@@ -15,7 +20,62 @@ func BasicRequestWithBody(c *cli.Context, method string) (string, error) {
+		}
+
+		color.Yellow("Waiting for file to close..\n\n")
+		cmd.Wait()
```
`cmd.Wait()` returns a pretty important error that is discarded here, namely if anything related to the execution of the command fails.

Reference: https://pkg.go.dev/os/exec#Cmd.Wait

Suggestion:
```go
if err := cmd.Wait(); err != nil {
    return "", fmt.Errorf("error in subprocess %s: %w", editor, err)
}
```

gbmor reviewed (commented):
Realized I left this out of my initial review, an important error return value was discarded.

athul reviewed (changes requested):
Hey @devblin the PR seems fine. Please address the changes and feedback given by @gbmor and me 🤝

athul commented on methods/basic.go:
```diff
@@ -15,7 +20,62 @@ func BasicRequestWithBody(c *cli.Context, method string) (string, error) {
+			return "", fmt.Errorf("Unable to open editor: %s", err)
+		}
+
+		color.Yellow("Waiting for file to close..\n\n")
```
Oh sorry, I forgot that we already use the color package. haha 😅  

So yeah, never mind this comment.

devblin commented:
@athul I have updated the changes, and I was getting some conflicts, I have resolved the conflicts in local with changes from **master** and I am getting error here: https://github.com/hoppscotch/hopp-cli/blob/8fc545289c46f5765aafe28c6372cc36b0d7194f/methods/basic.go#L29 as `undefined: io.ReadAll`, so should I updated it within this PR ?

gbmor commented:
@devblin change that to `ioutil.ReadAll(...)` ... it's related to my other PR which was to bring hopp-cli up to Go1.16, and that function didn't exist before 1.16

devblin commented:
> @devblin change that to `ioutil.ReadAll(...)` ... it's related to my other PR which was to bring hopp-cli up to Go1.16, and that function didn't exist before 1.16

I updated it within my last commit 👍🏽 .

athul commented:
@gbmor I request another round of review 🙈

athul commented:
@devblin please add the required changes to the readme too ✨

gbmor commented on methods/basic.go:
```diff
@@ -26,10 +29,61 @@ func BasicRequestWithBody(c *cli.Context, method string) (string, error) {
+			if currentOs == "linux" || currentOs == "darwin" {
+				editor = "nano"
+			} else if currentOs == "windows" {
+				editor = "notepad"
```
I don't have a windows machine to test this, and I haven't really used windows since 2006-2007 so I'm just guessing here: but will this still find notepad without the file extension, or should it be `notepad.exe`?

devblin replied:
@gbmor I have tested it on windows machine too.

gbmor replied:
@devblin OK cool! Good to know. I'll remember that for when I have to target Windows in the future :+1:

gbmor reviewed (commented):
Looks pretty much good to me, as far as the code goes. I had one question about calling `notepad` which I don't know the answer to because I don't have a windows machine to test it on, and I've only ever written software targeting Linux/OpenBSD/FreeBSD, but I tagged it in the review. @athul maybe you know? Or have a windows machine to test it on?

Also, maybe add a blurb in the readme about using `-e` and how it checks `$EDITOR` then falls back to `nano` or `notepad`?

devblin commented:
> @devblin please add the required changes to the readme too ✨

Updated README 👍🏽

athul commented on README.md:
```diff
@@ -146,3 +146,25 @@ $ cat myrequest.json
+
+### Providing a Request Body via text-editor
+
+In addition to providing request body via `-b/ --body` flag and stdin,
```
```suggestion
In addition to providing request body via `-b / --body` flag and stdin,
```

athul commented on README.md:
````diff
@@ -146,3 +146,25 @@ $ cat myrequest.json
+$ hopp-cli post https://reqres.in/api/users/2 -c js -e
+```
+
+It will preferrably open editor based on `$EDITOR` env.
````
```suggestion
It will preferrably open editor based on `$EDITOR` environment variable.
```

athul commented on README.md:
```diff
@@ -146,3 +146,25 @@ $ cat myrequest.json
+It will preferrably open editor based on `$EDITOR` env.
+
+**For example:**
+In your environment variable if `EDITOR=code` it will open VSCode for request-body input. Else, it will use default editor value based on OS.
```
```suggestion
If the environment variable is`$EDITOR=code` it will open VSCode for request-body input. Else, it will use default editor value based on the OS.
```

athul reviewed (approved):
Just some small stuff I think would be good, you can accept or ignore it completely. Anyways looks good to me. Sorry for taking forever to merge this. Was a bit tied up with some stuff

devblin commented:
I missed to mention that in Windows the print statement mentioned below wasn't properly color-encoding the `out` in terminal.
https://github.com/hoppscotch/hopp-cli/blob/8e853423d6170d3e573ea2442be3d785945f6dee/cli.go#L172
Instead, as mentioned in https://github.com/fatih/color/blob/master/doc.go#L89 , we should use 
`fmt.Fprintf(color.Output, out)`

athul commented:
> I missed to mention that in Windows the print statement mentioned below wasn't properly color-encoding the `out` in terminal.
> 
> https://github.com/hoppscotch/hopp-cli/blob/8e853423d6170d3e573ea2442be3d785945f6dee/cli.go#L172
> 
> 
> Instead, as mentioned in https://github.com/fatih/color/blob/master/doc.go#L89 , we should use
> `fmt.Fprintf(color.Output, out)`

Okay. Haven't had the chance to test everything on windows(I don't have a windows machine either). It'd be great if you could push that too

devblin commented:
Done

athul commented:
I guess this is ready to be merged. Thank you @devblin and @gbmor 🎉 🎉
